File: code/pole/pole/cpp/pole.cpp
//
// Created by elvieto on 23-11-18.
//

#include "pole.h"
#include <cmath>

#define MAX_THETA 3.141592653589793 // 0.523598776  // 30 degrees

const double pi = 3.141592653589793;
const double pi2 = 6.283185307179586;

Pole::Pole(double mass, double length):
        theta(0),
        theta_dot(0),
        time(0),
        dt(0.01),
        length(length),
        mass(mass),
        inertia(mass * length * length / 3){

}

void Pole::act(double torque, double &reward, bool &is_terminal) {
//    torque = 0;
    torque += mass * 9.81 * (length / 2) * sin(theta);
    // Rotational acceleration
    double theta_dot_dot = torque / inertia;

    // Integrate
    theta_dot += theta_dot_dot * dt;
    theta += theta_dot * dt;
    time += dt;

    theta = fmod(fmod(theta + pi, pi2) + pi2, pi2) - pi;

    reward = pow((MAX_THETA - fabs(theta)) / MAX_THETA, 4) - 0.2;
    is_terminal = false;

//    if (fabs(theta) > MAX_THETA) {
//        reward = -1.0;
//        is_terminal = true;
//    } else {
//        reward = pow((MAX_THETA - fabs(theta)) / MAX_THETA, 16) - 0.2;
//        is_terminal = false;
//    }
}

File: code/pypole/pypole/pole.py
from math import sin, fmod
from typing import Optional

import numpy as np

from cw.simulation import ModuleBase
from cw.constants import g_earth, pi2, pi


class Pole(ModuleBase):
    def __init__(self,
                 mass: float,
                 length: float,
                 theta_terminate: Optional[float]=None,
                 theta_min_reward: Optional[float]=None):
        super().__init__(
            required_states=[
                "agent_torque",
                "theta",
                "theta_dot",
                "theta_dot_dot",
                "torque",
                "reward"
            ]
        )

        self.mass = mass
        self.length = length
        self.theta_terminate = theta_terminate
        self.theta_min_reward = theta_min_reward or pi
        self.state = np.array([0.0, 0.0])
        self.inertia = mass * length * length / 3

    def step(self):
        theta = self.s.theta
        abs_theta = abs(theta)
        
        self.s.torque = self.mass * g_earth * (self.length / 2) * sin(theta) + self.s.agent_torque
        self.s.theta_dot_dot = self.s.torque / self.inertia
        self.s.theta = fmod(fmod(theta + pi, pi2) + pi2, pi2) - pi
        self.s.reward = ((self.theta_min_reward - abs(theta)) / self.theta_min_reward) ** 4

        if self.theta_terminate:
            if abs_theta > self.theta_terminate:
                self.simulation.stop()
